### PR TITLE
[Merged by Bors] - feat: Small sets of games/surreals are bounded

### PR DESCRIPTION
**Pull request by YaelDillies**
Finish forwarding porting https://github.com/leanprover-community/mathlib/pull/15260 after #10566 ported just the changes in `PGame.lean`.

We don't port `upper_bound_numeric` (or `lower_bound_numeric`) because `upper_bound` was deleted following review feedback on the previous forward-porting PR #10566.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Normed/Order/UpperLower.lean
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/Normed/Order/UpperLower.lean:128]:
```suggestion
-- TODO: Generalise those lemmas so that they also apply to `ℝ` and `EuclideanSpace ι ℝ`
```

File: Mathlib/Topology/Algebra/Order/UpperLower.lean
Comment (urkud) [inline Mathlib/Topology/Algebra/Order/UpperLower.lean:88]: `BddBelow.of_closure` doesn't need any assumptions on the relation of the topology and the order. Please prove it without assumptions.

File: Mathlib/Topology/Algebra/Order/UpperLower.lean
Comment (urkud) [inline Mathlib/Topology/Algebra/Order/UpperLower.lean:70]: Please move this to the recently created `Topology/Order/OrderClosed` and assume `ClosedIicTopology` instead of `OrderClosedTopology`.

**Comment by urkud** [comment]
Could you please split it into smaller PRs, each for 1 topic?

**Comment by YaelDillies** [comment]
It's a bit of a pain to split, so I'll only do it if you have a specific thing you would like to see in another PR.

File: Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean
Comment (urkud) [inline Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean:1909]: Can you reuse some `variable`s here? Also, I think that `g` should be implicit.

File: Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean
Comment (urkud) [inline Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean:1910]:
```suggestion
    AEStronglyMeasurable g ν := let ⟨g₁, hg₁, hg₁'⟩ := hμ; ⟨g₁, hg₁, h.ae_eq hg₁'⟩
```

**Comment by urkud** [comment]
> It's a bit of a pain to split, so I'll only do it if you have a specific thing you would like to see in another PR.

Is it a pain to split this particular PR or you lack tools to quickly split a PR in general? In the latter case, I can split this PR for you.

File: Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean
Comment (urkud) [inline Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean:1909]: Loogle says this already exists as [mono'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.html#MeasureTheory.AEStronglyMeasurable.mono%27). We can add an `alias` though.

**Comment by YaelDillies** [comment]
Indeed I don't know how to cherry-pick changes efficiently. I've now opened #10556 from your branch.

File: Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean
Comment (YaelDillies) [reply Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean]: `mono_ac` is the name I was looking for. I renamed it now.

**Comment by urkud** [comment]
Emacs magit allows me to look at the diff with another branch, then apply chunks from this diff with a single keystroke. In the command line you can `git reset --soft other-branch`, then `git add -i` but this is less convenient. Not sure about VS Code integration.

**Comment by urkud** [comment]
I split this PR into 3 branches:

- https://github.com/leanprover-community/mathlib4/tree/YK-Yael-pgame
- https://github.com/leanprover-community/mathlib4/tree/YK-Yael-normed-order
- https://github.com/leanprover-community/mathlib4/tree/YK-Yael-mono_ac

I will merge the `mono_ac` PR as soon as you open it. I can review the `normed-order` PR but I have to read the definitions in that file first. I know nothing about `PGame`s, so I won't review that PR.

**Comment by eric-wieser** [comment]
@urkud: it sounds like you're taking the stance that we are not bothering to check if these are faithful forward ports. I think that's probably defensible, but we should make sure to record that choice in each commit message.

**Comment by urkud** [comment]
I think that at this moment Mathlib 3 and Mathlib 4 already diverged a lot. I found a few issues in this PR (coming from mathlib3 version). IMHO, even if we merged imperfect code into Mathlib 3, this doesn't force us to merge it in Mathlib 4 now.

I agree that we should mention that the code was improved while forward-porting (too late for #10556).

File: Mathlib/Analysis/Normed/Order/UpperLower.lean
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/Normed/Order/UpperLower.lean:67]: Is the additive name correct?

Comment (YaelDillies) [reply Mathlib/Analysis/Normed/Order/UpperLower.lean]: No. Well spotted.

File: Mathlib/Analysis/Normed/Order/UpperLower.lean
Comment (eric-wieser) [inline Mathlib/Analysis/Normed/Order/UpperLower.lean:245]: `is_antichain.interior_eq_empty` seems to be missing from the forward-port here; if this is deliberate, there should be a `noalign`

**Review by eric-wieser — DISMISSED** [review]
I updated the out-of-sync dashboard; these are not the final two forward-ports, but they are *your* final two forward-ports!

I see one lemma missing, but otherwise these look like faithful ports.

bors d+

**Comment by eric-wieser** [comment]
> I agree that we should mention that the code was improved while forward-porting

Please also modify the description to do this before merging, if anything has been improved.

File: Mathlib/SetTheory/Surreal/Basic.lean
Comment (YaelDillies) [inline Mathlib/SetTheory/Surreal/Basic.lean:406]: @timotree3, your suggestion of not having a separate definition for `upperBound` is backfiring. Would you mind having a look (and fixing it because I already spent too much time on this 😇)

Comment (timotree3) [reply Mathlib/SetTheory/Surreal/Basic.lean]: Indeed. We can't examine the witness of `PGame.bddAbove_range_of_small` and see that it's numeric anymore. For now I've just duplicated the proof. I think it would be cleaner to prove `Surreal.bddAbove_range_of_small` through transitivity with `PGame.bddAbove_range_of_small` and a new lemma `Numeric.exists_ge`, but I think that's too big to do in this PR.

(Here is the cleanup I'm picturing eventually)
```lean
@[simp] lemma _root_.SetTheory.Game.mk_le_mk {x y : PGame.{u}} : (⟦x⟧ : Game) ≤ ⟦y⟧ ↔ x ≤ y :=
  Iff.rfl

@[simp] lemma mk_le_mk {x y : PGame.{u}} {hx hy}: mk x hx ≤ mk y hy ↔ x ≤ y := Iff.rfl

@[elab_as_elim]
theorem inductionOn (x : Surreal) {P : Surreal → Prop} (h : ∀ y h, P (mk y h)) : P x := by
  induction' x using Quotient.inductionOn with x
  rcases x with ⟨x, hx⟩
  apply h

lemma _root_.SetTheory.PGame.Numeric.exists_ge (x : PGame) : ∃ y, x ≤ y ∧ y.Numeric := sorry

/-- A small family of surreals is bounded above. -/
lemma bddAbove_range_of_small {ι : Type*} [Small.{u} ι] (f : ι → Surreal.{u}) :
    BddAbove (Set.range f) := by
  let g : ι → Game.{u} := toGame ∘ f
  have ⟨x, hx⟩ := Game.bddAbove_range_of_small g
  induction' x using Quotient.ind with x
  have ⟨y, hxy, hy⟩ := Numeric.exists_ge x
  refine ⟨mk y hy, fun z ⟨i, hfi⟩ ↦ ?_⟩
  induction' z using inductionOn with z hz
  specialize @hx ⟦z⟧ ⟨i, by simp [hfi]⟩
  simp only [Game.mk_le_mk, mk_le_mk] at hx ⊢
  exact hx.trans hxy
```

**Comment by timotree3** [comment]
This PR is in a confusing state. It looks like when Eric delegated, the PR was about Analysis/Normed/Order/UpperLower and now it's about SetTheory/Surreal. I just pushed a commit finishing a proof in SetTheory/Surreal that Yaël asked me to but I'm sure there will be requested changes to the proof because I haven't submitted many to mathlib before. Should this be marked `awaiting-review` again?

**Review by timotree3 — APPROVED** [review]
I approve of the changes in Mathlib/SetTheory/Game/Basic.lean. The changes in Mathlib/SetTheory/Surreal/Basic.lean are basically my own now so need another pair of eyes

**Comment by YaelDillies** [comment]
This PR is and always has been about the final forward-ports. However what counts as "final forward-ports" changed since Yury asked me to split a bunch of things off to their own PRs and Eric found new files that were out of sync.

File: Mathlib/SetTheory/Surreal/Basic.lean
Comment (YaelDillies) [inline Mathlib/SetTheory/Surreal/Basic.lean:326]:
```suggestion
lemma mk_le_mk {x y : PGame.{u}} {hx hy} : mk x hx ≤ mk y hy ↔ x ≤ y := Iff.rfl
```

File: Mathlib/SetTheory/Surreal/Basic.lean
Comment (YaelDillies) [inline Mathlib/SetTheory/Surreal/Basic.lean:431]:
```suggestion
/-- A small set of surreals is bounded below. -/
```

File: Mathlib/SetTheory/Game/Basic.lean
Comment (eric-wieser) [inline Mathlib/SetTheory/Game/Basic.lean:209]:
```suggestion
  simpa [PGame.le_iff_game_le] using hx <| Set.mem_range_self i
```
and below

File: Mathlib/SetTheory/Surreal/Basic.lean
Comment (eric-wieser) [inline Mathlib/SetTheory/Surreal/Basic.lean:391]: What happened to `upperBound_numeric`?

**Review by eric-wieser — COMMENTED** [review]
bors d-

(since the version I delegated was different)

File: Mathlib/SetTheory/Surreal/Basic.lean
Comment (timotree3) [reply Mathlib/SetTheory/Surreal/Basic.lean:391]: In the review for #10566 we decided to get rid of the definition `upperBound` entirely just keep `PGame.bddAbove_of_small`. I don't think we realized that this file relied on specific properties of the witness used. For now, I've duplicated the proof of `PGame.bddAbove_of_small`, but in the future I think long term this proof should be refactored to take the (possibly non-numeric) upper bound given `PGame.bddAbove_of_small` and then use the fact that there is a `Numeric` game greater than any given game to provide a `Numeric` upper bound. 

(I also wrote this in an earlier comment in this PR https://github.com/leanprover-community/mathlib4/pull/10458#discussion_r1510020804)

Comment (timotree3) [reply Mathlib/SetTheory/Surreal/Basic.lean:391]: Oh, are you just saying there needs to be a `#noalign`?

Comment (eric-wieser) [reply Mathlib/SetTheory/Surreal/Basic.lean:391]: It should either be present, or `noalign` along with an explanation of its absence living in the PR description.

Comment (timotree3) [reply Mathlib/SetTheory/Surreal/Basic.lean:391]: I added `#noalign`s

**Review by eric-wieser — APPROVED** [review]
bors merge

Thanks!

**Comment by eric-wieser** [comment]
bors r-

bors d+

Looks like an easy fix.

**Comment by YaelDillies** [comment]
bors merge

**Comment by YaelDillies** [comment]
bors merge

**Pull request closed without merging**
